File: sora-2-api-asynchronous/image-or-text-to-video/sora-2-async.js
/**
 * Sora-2 视频生成完整异步流程 - JavaScript/Node.js 版本
 * 智能双模式：自动识别文字生成视频 / 图片生成视频
 *
 * 依赖安装: npm install axios form-data
 * 运行: node sora-2-async.js
 */

const axios = require('axios');
const FormData = require('form-data');
const fs = require('fs');
const path = require('path');

// ============================================
// 配置参数
// ============================================
const CONFIG = {
    baseURL: 'https://api.apiyi.com/v1/videos',
    apiKey: 'sk-',

    // ==================== 模式配置 ====================
    // 设置 imagePath 来切换生成模式：
    // - imagePath: null (或留空)    → 文字生成视频
    // - imagePath: '/path/to/image' → 图片生成视频
    // =================================================

    imagePath: null,  // 文字生成视频模式
    // imagePath: '/path/to/your/image.png',  // 取消注释并设置路径，切换为图片生成视频模式

    prompt: '一只橘色的小猫在阳光明媚的花园里追逐蝴蝶',
    model: 'sora-2',
    size: '1280x720',
    seconds: '15',
    pollInterval: 30000,  // 30 秒
    maxWaitTime: 600000,  // 10 分钟
};

/**
 * 根据文件扩展名推断 MIME 类型
 */
function getMimeType(filePath) {
    const ext = path.extname(filePath).toLowerCase();
    const mimeTypes = {
        '.png': 'image/png',
        '.jpg': 'image/jpeg',
        '.jpeg': 'image/jpeg',
    };
    return mimeTypes[ext] || 'image/png';
}

/**
 * 格式化时间戳
 */
function formatTimestamp(timestamp) {
    const date = new Date(timestamp * 1000);
    return date.toLocaleString('zh-CN', {
        year: 'numeric',
        month: '2-digit',
        day: '2-digit',
        hour: '2-digit',
        minute: '2-digit',
        second: '2-digit',
    });
}

/**
 * 格式化文件大小
 */
function formatFileSize(bytes) {
    if (bytes < 1024) return bytes + ' B';
    if (bytes < 1024 * 1024) return (bytes / 1024).toFixed(2) + ' KB';
    return (bytes / (1024 * 1024)).toFixed(2) + ' MB';
}

/**
 * 第一步：提交视频生成请求（自动识别文字或图片模式）
 */
async function step1SubmitRequest() {
    console.log('============================================================');
    console.log('第一步：提交视频生成请求');
    console.log('============================================================');

    // 检查是否为图片生成模式
    const isImageMode = CONFIG.imagePath && fs.existsSync(CONFIG.imagePath);

    if (CONFIG.imagePath && !isImageMode) {
        console.log(`⚠️  警告：图片路径已设置但文件不存在: ${CONFIG.imagePath}`);
        console.log(`    将使用文字生成视频模式`);
    }

    // 构建 FormData
    const formData = new FormData();
    formData.append('prompt', CONFIG.prompt);
    formData.append('model', CONFIG.model);
    formData.append('size', CONFIG.size);
    formData.append('seconds', CONFIG.seconds);

    try {
        if (isImageMode) {
            // 图片生成视频模式
            console.log('📷 模式：图片生成视频');
            console.log('📤 发送请求...');
            console.log(`   - 提示词: ${CONFIG.prompt}`);
            console.log(`   - 图片: ${CONFIG.imagePath}`);
            console.log(`   - 模型: ${CONFIG.model}`);
            console.log(`   - 尺寸: ${CONFIG.size}`);
            console.log(`   - 时长: ${CONFIG.seconds}秒`);

            formData.append('input_reference', fs.createReadStream(CONFIG.imagePath), {
                filename: path.basename(CONFIG.imagePath),
                contentType: getMimeType(CONFIG.imagePath),
            });
        } else {
            // 文字生成视频模式
            console.log('📝 模式：文字生成视频');
            console.log('📤 发送请求...');
            console.log(`   - 提示词: ${CONFIG.prompt}`);
            console.log(`   - 模型: ${CONFIG.model}`);
            console.log(`   - 尺寸: ${CONFIG.size}`);
            console.log(`   - 时长: ${CONFIG.seconds}秒`);
        }

        const response = await axios.post(CONFIG.baseURL, formData, {
            headers: {
                ...formData.getHeaders(),
                'Authorization': CONFIG.apiKey,
            },
        });

        const result = response.data;
        console.log('✅ 请求提交成功!');
        console.log(`   - 视频ID: ${result.id}`);
        console.log(`   - 状态: ${result.status}`);
        console.log(`   - 创建时间: ${formatTimestamp(result.created_at)}`);
        console.log(`   ⏱️  预计生成时间: 3-5 分钟`);

        return { videoId: result.id, isImageMode };
    } catch (error) {
        if (error.response) {
            throw new Error(`❌ 请求失败: ${error.response.status}\n响应: ${JSON.stringify(error.response.data)}`);
        }
        throw new Error(`❌ 请求失败: ${error.message}`);
    }
}

/**
 * 第二步：轮询查询视频状态
 */
async function step2PollStatus(videoId) {
    console.log('\n============================================================');
    console.log('第二步：查询视频生成状态');
    console.log('============================================================');
    console.log(`💡 提示：视频生成通常需要 3-5 分钟，每 ${CONFIG.pollInterval / 1000} 秒自动查询一次`);

    const startTime = Date.now();

    while (true) {
        const elapsed = Date.now() - startTime;

        // 检查超时
        if (elapsed > CONFIG.maxWaitTime) {
            throw new Error(`⏱️ 超时：已等待 ${CONFIG.maxWaitTime / 1000} 秒，停止查询`);
        }

        console.log(`\n🔍 查询状态... (已等待 ${Math.floor(elapsed / 1000)} 秒)`);

        try {
            const response = await axios.get(`${CONFIG.baseURL}/${videoId}`, {
                headers: {
                    'Authorization': CONFIG.apiKey,
                },
            });

            const result = response.data;
            console.log(`   - 状态: ${result.status}`);
            console.log(`   - 进度: ${result.progress || 0}%`);

            switch (result.status) {
                case 'completed':
                    console.log('✅ 视频生成完成!');
                    console.log(`   - 视频URL: ${result.url}`);
                    console.log(`   - 完成时间: ${formatTimestamp(result.completed_at)}`);
                    return;
                case 'failed':
                    throw new Error('❌ 视频生成失败');
                case 'submitted':
                case 'in_progress':
                    console.log(`⏳ 视频生成中，等待 ${CONFIG.pollInterval / 1000} 秒后继续查询...`);
                    await sleep(CONFIG.pollInterval);
                    break;
                default:
                    console.log(`⚠️ 未知状态: ${result.status}`);
                    await sleep(CONFIG.pollInterval);
            }
        } catch (error) {
            if (error.response) {
                console.log(`❌ 状态查询失败: ${error.response.status}`);
            } else if (error.message.includes('视频生成失败')) {
                throw error;
            } else {
                console.log(`❌ 查询发生异常: ${error.message}`);
            }
            await sleep(CONFIG.pollInterval);
        }
    }
}

/**
 * 第三步：下载视频
 */
async function step3DownloadVideo(videoId, isImageMode) {
    console.log('\n============================================================');
    console.log('第三步：下载视频');
    console.log('============================================================');

    // 根据模式生成文件名
    const modePrefix = isImageMode ? 'image' : 'text';
    const outputFile = `sora_${modePrefix}_video_${videoId.slice(-8)}.mp4`;

    console.log('📥 开始下载视频...');

    try {
        const response = await axios.get(`${CONFIG.baseURL}/${videoId}/content`, {
            headers: {
                'Authorization': CONFIG.apiKey,
            },
            responseType: 'stream',
        });

        const writer = fs.createWriteStream(outputFile);
        let downloadedBytes = 0;
        const totalBytes = parseInt(response.headers['content-length'] || '0');

        response.data.on('data', (chunk) => {
            downloadedBytes += chunk.length;
            if (totalBytes > 0) {
                const percent = ((downloadedBytes / totalBytes) * 100).toFixed(1);
                process.stdout.write(`\r   下载进度: ${percent}% (${downloadedBytes}/${totalBytes} 字节)`);
            }
        });

        response.data.pipe(writer);

        return new Promise((resolve, reject) => {
            writer.on('finish', () => {
                const stats = fs.statSync(outputFile);
                console.log('\n✅ 视频下载成功!');
                console.log(`   - 保存路径: ${path.resolve(outputFile)}`);
                console.log(`   - 文件大小: ${formatFileSize(stats.size)}`);
                resolve();
            });
            writer.on('error', reject);
        });
    } catch (error) {
        if (error.response) {
            throw new Error(`❌ 下载失败: ${error.response.status}`);
        }
        throw new Error(`❌ 下载失败: ${error.message}`);
    }
}

/**
 * 辅助函数：延迟
 */
function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

/**
 * 主流程
 */
async function main() {
    const startTime = new Date();

    console.log('\n🎬 开始 Sora-2 视频生成完整流程');
    console.log(`⏰ 开始时间: ${startTime.toLocaleString('zh-CN')}\n`);

    try {
        // 第一步：提交请求
        const { videoId, isImageMode } = await step1SubmitRequest();

        // 第二步：轮询状态
        await step2PollStatus(videoId);

        // 第三步：下载视频
        await step3DownloadVideo(videoId, isImageMode);

        console.log('\n============================================================');
        console.log('🎉 完整流程执行成功!');
        console.log('============================================================');
        console.log(`⏰ 结束时间: ${new Date().toLocaleString('zh-CN')}`);
    } catch (error) {
        console.error(`\n❌ 流程终止：${error.message}`);
        process.exit(1);
    }
}

// 运行主流程
if (require.main === module) {
    main();
}

module.exports = { main };

File: sora-2-api-asynchronous/image-or-text-to-video/sora-2-async.py
import requests
import time
import os
from datetime import datetime

# API 配置
BASE_URL = "https://api.apiyi.com/v1/videos"
API_KEY = "sk-"

# ==================== 模式配置 ====================
# 设置 IMAGE_PATH 来切换生成模式：
# - IMAGE_PATH = None (或留空)    → 文字生成视频
# - IMAGE_PATH = '/path/to/image' → 图片生成视频
# =================================================

#IMAGE_PATH = None  # 取消注释，即文字生成视频模式，注释即图片生成视频模式；本文件默认是图片生成视频模式
IMAGE_PATH = r'./dog-and-cat.png'  # 取消注释并设置路径，切换为图片生成视频模式

# 请求参数配置
PROMPT = '参考配图，使得动物们活跃起来，在场地里绕场一周的追逐，画面要生动有趣，背景音乐类似：让我们欢快地跳起来'
MODEL = 'sora-2'
SIZE = '1280x720'  # 或 '720x1280'
SECONDS = '15'     # 支持 '10' 或 '15'

# 轮询配置
POLL_INTERVAL = 30  # 每 30 秒查询一次状态
MAX_WAIT_TIME = 600  # 最长等待 10 分钟


def get_mime_type(file_path):
    """根据文件扩展名推断 MIME 类型"""
    ext = os.path.splitext(file_path)[1].lower()
    mime_types = {
        '.png': 'image/png',
        '.jpg': 'image/jpeg',
        '.jpeg': 'image/jpeg',
    }
    return mime_types.get(ext, 'image/png')


def step1_submit_video_request():
    """第一步：提交视频生成请求（自动识别文字或图片模式）"""
    print("=" * 60)
    print("第一步：提交视频生成请求")
    print("=" * 60)

    # 检查是否为图片生成模式
    is_image_mode = IMAGE_PATH and os.path.exists(IMAGE_PATH)

    if IMAGE_PATH and not is_image_mode:
        print(f"⚠️  警告：图片路径已设置但文件不存在: {IMAGE_PATH}")
        print(f"    将使用文字生成视频模式")

    # 基础参数（使用 multipart/form-data 格式）
    files = {
        'prompt': (None, PROMPT),
        'model': (None, MODEL),
        'size': (None, SIZE),
        'seconds': (None, SECONDS)
    }

    headers = {
        'Accept': 'application/json',
        'Authorization': API_KEY,
        'Host': 'api.apiyi.com',
        'Connection': 'keep-alive'
    }

    try:
        if is_image_mode:
            # 图片生成视频模式
            print("📷 模式：图片生成视频")
            print(f"📤 发送请求...")
            print(f"   - 提示词: {PROMPT}")
            print(f"   - 图片: {IMAGE_PATH}")
            print(f"   - 模型: {MODEL}")
            print(f"   - 尺寸: {SIZE}")
            print(f"   - 时长: {SECONDS}秒")

            with open(IMAGE_PATH, 'rb') as f:
                files['input_reference'] = (
                    os.path.basename(IMAGE_PATH),
                    f,
                    get_mime_type(IMAGE_PATH)
                )
                response = requests.post(
                    BASE_URL,
                    headers=headers,
                    files=files
                )
        else:
            # 文字生成视频模式
            print("📝 模式：文字生成视频")
            print(f"📤 发送请求...")
            print(f"   - 提示词: {PROMPT}")
            print(f"   - 模型: {MODEL}")
            print(f"   - 尺寸: {SIZE}")
            print(f"   - 时长: {SECONDS}秒")

            response = requests.post(
                BASE_URL,
                headers=headers,
                files=files
            )

        if response.status_code == 200:
            result = response.json()
            video_id = result.get('id')
            print(f"✅ 请求提交成功!")
            print(f"   - 视频ID: {video_id}")
            print(f"   - 状态: {result.get('status')}")
            print(f"   - 创建时间: {datetime.fromtimestamp(result.get('created_at'))}")
            print(f"   ⏱️  预计生成时间: 3-5 分钟")
            return video_id, is_image_mode
        else:
            print(f"❌ 请求失败: {response.status_code}")
            print(f"   响应: {response.text}")
            return None, None

    except Exception as e:
        print(f"❌ 请求发生异常: {str(e)}")
        return None, None


def step2_poll_video_status(video_id):
    """第二步：轮询查询视频生成状态"""
    print("\n" + "=" * 60)
    print("第二步：查询视频生成状态")
    print("=" * 60)
    print(f"💡 提示：视频生成通常需要 3-5 分钟，每 {POLL_INTERVAL} 秒自动查询一次")

    headers = {
        'Authorization': API_KEY
    }

    status_url = f"{BASE_URL}/{video_id}"
    start_time = time.time()

    while True:
        elapsed_time = time.time() - start_time

        # 检查是否超时
        if elapsed_time > MAX_WAIT_TIME:
            print(f"⏱️ 超时：已等待 {MAX_WAIT_TIME} 秒，停止查询")
            return False

        try:
            print(f"\n🔍 查询状态... (已等待 {int(elapsed_time)} 秒)")
            response = requests.get(status_url, headers=headers)

            if response.status_code == 200:
                result = response.json()
                status = result.get('status')
                progress = result.get('progress', 0)

                print(f"   - 状态: {status}")
                print(f"   - 进度: {progress}%")

                if status == 'completed':
                    print(f"✅ 视频生成完成!")
                    print(f"   - 视频URL: {result.get('url')}")
                    print(f"   - 完成时间: {datetime.fromtimestamp(result.get('completed_at'))}")
                    return True
                elif status == 'failed':
                    print(f"❌ 视频生成失败")
                    return False
                elif status in ['submitted', 'in_progress']:
                    print(f"⏳ 视频生成中，等待 {POLL_INTERVAL} 秒后继续查询...")
                    time.sleep(POLL_INTERVAL)
                else:
                    print(f"⚠️ 未知状态: {status}")
                    time.sleep(POLL_INTERVAL)
            else:
                print(f"❌ 状态查询失败: {response.status_code}")
                print(f"   响应: {response.text}")
                time.sleep(POLL_INTERVAL)

        except Exception as e:
            print(f"❌ 查询发生异常: {str(e)}")
            time.sleep(POLL_INTERVAL)


def step3_download_video(video_id, is_image_mode, output_path=None):
    """第三步：下载生成的视频"""
    print("\n" + "=" * 60)
    print("第三步：下载视频")
    print("=" * 60)

    # 根据模式生成默认文件名
    if output_path is None:
        mode_prefix = "image" if is_image_mode else "text"
        output_path = f"sora_{mode_prefix}_video_{video_id.split('_')[-1]}.mp4"

    headers = {
        'Authorization': API_KEY
    }

    content_url = f"{BASE_URL}/{video_id}/content"

    try:
        print(f"📥 开始下载视频...")
        response = requests.get(content_url, headers=headers, stream=True)

        if response.status_code == 200:
            # 获取文件大小
            total_size = int(response.headers.get('content-length', 0))

            with open(output_path, 'wb') as f:
                if total_size == 0:
                    f.write(response.content)
                else:
                    downloaded = 0
                    for chunk in response.iter_content(chunk_size=8192):
                        if chunk:
                            f.write(chunk)
                            downloaded += len(chunk)
                            percent = (downloaded / total_size) * 100
                            print(f"\r   下载进度: {percent:.1f}% ({downloaded}/{total_size} 字节)", end='')

            print(f"\n✅ 视频下载成功!")
            print(f"   - 保存路径: {os.path.abspath(output_path)}")
            print(f"   - 文件大小: {os.path.getsize(output_path)} 字节")
            return True
        else:
            print(f"❌ 下载失败: {response.status_code}")
            print(f"   响应: {response.text}")
            return False

    except Exception as e:
        print(f"❌ 下载发生异常: {str(e)}")
        return False


def main():
    """完整的异步流程"""
    print("\n🎬 开始 Sora-2 视频生成完整流程")
    print(f"⏰ 开始时间: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}\n")

    # 第一步：提交请求
    video_id, is_image_mode = step1_submit_video_request()
    if not video_id:
        print("\n❌ 流程终止：视频请求提交失败")
        return

    # 第二步：轮询状态
    success = step2_poll_video_status(video_id)
    if not success:
        print("\n❌ 流程终止：视频生成未完成")
        return

    # 第三步：下载视频
    success = step3_download_video(video_id, is_image_mode)

    if success:
        print("\n" + "=" * 60)
        print("🎉 完整流程执行成功!")
        print("=" * 60)
        print(f"⏰ 结束时间: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}")
    else:
        print("\n❌ 流程终止：视频下载失败")


if __name__ == "__main__":
    main()
